### DISCUSSION TITLE
Having trouble with the led blinker tutorial using the raspberry 3 model b+ when sending "led.BLINKING_ON_OFF" command

### DISCUSSION THREAD

**Discussion by tbkevin** [nasa/fprime#2797] [Q&A] [answered]
Hi everyone. My name is Kevin and I've started the tutorials for an ongoing project i have. For the deployment in the raspberry of the led blinker tutorial, i'm facing the next issue related, i think, with the gpio linux driver. I sucessfully built the binaries needed for the raspberry pi in my host computer, running a Ubuntu 22.04, and upladed it onto the raspberry. When i execute both gds and binaries everything works fine, but when i send the command "led.BLINKING_ON_OFF" to start blinking the led, the software in the raspberry crash and log the next message: 
![image](https://github.com/nasa/fprime/assets/36642912/3b22bcfa-4135-4c22-9fb9-880c3d312154)
And the line of code that throw this error is the next one but i'm not able of figure it out what problem is:
![image](https://github.com/nasa/fprime/assets/36642912/a2f50023-fde4-4d69-a837-99da853d59e6)

It doesn't happen if i send it with the OFF argument.

I would aprreciate any hint to tackle this issue.

Thanks!

**Comment by Joshua-Anderson** [comment]
Do you see any log messages about failing to open the GPIO pin?

You may need to run the flight software as root to have GPIO pin permissions

**Reply by tbkevin** [reply]
Hi, Joshua. I have checked that i got the gpio permissions using the 'groups' command. Also always ran the software on the raspberry using 'sudo'. Just to be sure, the commands i use on the host computer for the gds and on the raspberry are these:
"fprime-gds -n --dictionary build-artifacts/aarch64-linux/LedBlinker/dict/LedBlinkerTopologyAppDictionary.xml --ip-client --ip-address 192.168.1.19" and "pi@raspberrypi:~ $ sudo ledblinker/bin/LedBlinker -a 0.0.0.0 -p 50000"
Thare are no more logs after the 'Aborted', the program ends. Neither in the gds console.

Thanks for the reply, Joshua!

**Reply by Joshua-Anderson** [reply]
Did you get any log messages from the flight software about failing to open the GPIO pin?

**Comment by jhsu3** [comment, marked as answer]
I wrote an answer on an issue in the led-blinker repository [https://github.com/fprime-community/fprime-workshop-led-blinker/issues/67](url)

You might have a GPIO open error further up in the log where the component is initialized. If this is the case, this answer should help you.
Fprime is using sysfs to open the GPIO pins. Looking in `/sys/class/gpio` on my Pi4 shows gpiochip512 and gpiochip570.
The new addressing uses these chip### as the base for pin addresses. 

As an example, to open pin 6, your deployment xxxTopology.cpp file would have:
`bool gpio_success = gpioDriver.open((512+6), Drv::LinuxGpioDriver::GpioDirection::GPIO_OUT);`

The LinuxGpioDriver might need an update to accommodate the new sysfs or the tutorial can be updated to reflect the pin addressing change.

**Reply by Joshua-Anderson** [reply]
Yeah, the linux kernel is encouraging folks to move towards using the gpio chip block interface and libgpiod over the sysfs interface.  Fprime should definitely support a libgpiod driver in the future.

I think the raspberry pi linux kernels still supports sysfs, though maybe something changed there recently

**Reply by tbkevin** [reply]
Oh thanks a lot, it worked.Thank both of you for the help. Also, i found that the file "/sys/kernel/debug" contain the proper values of the gpios that i should use in the software.

Thanks a lot!
